opengovsg/GoGovSG
# Experiment go <> sgid integration

## Problem

We are trying to integrate SGID with GO and will be testing it with OGP officers. If the integration is successful, we will open this up to other public officers as well. 


## Solution

Integration doc with the detailed solution [here](https://www.notion.so/opengov/Go-SGID-Integration-Working-Doc-9a5428038168411ba877a25827d867df), for summary refer below:

Followed developer docs [here](https://docs.id.gov.sg/introduction/overview) 

1. Add a new login page for SGID login `/ogp-login`
2. Upon user click we redirect to SGID authentication url (opens up in the same window)
3. Once user allows sign in using GO SGID client, we fetch their relevant data i.e. officer's work email.
4. There's a validation check on valid officer's email ending in `.gov.sg`
5. If the validation succeeds we create or find the user in the DB and set the user session
6. Else we redirect the back to sgid login page and show the relevant error message.

**Features**:
Login feature with SGID

## Before & After Screenshots

**AFTER**:

https://github.com/opengovsg/GoGovSG/assets/25703976/ad4cf77c-f29e-4242-b601-ed58fbcaf555

Tested on health and edu domains, we get a 400 response, with this error message on the UI: 

<img width="1512" alt="Screenshot 2023-08-03 at 5 59 32 PM" src="https://github.com/opengovsg/GoGovSG/assets/25703976/82002e87-96e7-43a7-ab47-ddb2d7433f14">


## Tests

_What tests should be run to confirm functionality?_

1. Check login using SGID manually
2. Check remaining functionality remains intact [regression testing]
3. Check invalid authentication or invalid email does not succeed login
4. Check that edu and health domains are still intact, without the SGID client being initialised

## Deploy Notes
Need to create the relevant environment variables before deploying. 
There might be a need to use the AWS key manager, in case env variable store in EB does not work. 

**New environment variables**:

- SGID Client 
- SGID Private key 
- SGID API hostname
- SGID Client Secret
